lib/rspec/matchers/expecteds_for_multiple_diffs.rb

```diff
@@ -0,0 +1,75 @@
+module RSpec
+  module Matchers
+    # @api private
+    # Handles list of expected values when there is a need to render
+    # multiple diffs. Also can handle one value.
+    class ExpectedsForMultipleDiffs
+      # @private
+      # Default diff label when there is only one matcher in diff
+      # output
+      DEFAULT_DIFF_LABEL = "Diff:".freeze
+
+      # @private
+      # Maximum readable matcher description length
+      DESCRIPTION_MAX_LENGTH = 65
+
+      def initialize(expected_list)
+        @expected_list = expected_list
+      end
+
+      # @api private
+      # Wraps provided expected value in instance of
+      # ExpectedForMultipleDiffs. If provided value is already an
+      # ExpectedForMultipleDiffs then it just returns it.
+      # @param [Any] expected value to be wrapped
+      # @return [RSpec::Matchers::ExpectedsForMultipleDiffs]
+      def self.from(expected)
+        return expected if self === expected
+        new([[expected, DEFAULT_DIFF_LABEL]])
+      end
+
+      # @api private
+      # Wraps provided matcher list in instance of
+      # ExpectedForMultipleDiffs.
+      # @param [Array<Any>] matchers list of matchers to wrap
+      # @return [RSpec::Matchers::ExpectedsForMultipleDiffs]
+      def self.for_many_matchers(matchers)
+        new(matchers.map { |m| [m.expected, diff_label_for(m)] })
+      end
+
+      # @api private
+      # Returns message with diff(s) appended for provided differ
+      # factory and actual value if there are any
+      # @param [String] message original failure message
+      # @param [Proc] differ_factory brand new differ factory method (or proc, block)
+      # @param [Any] actual value
+      # @return [String]
+      def message_with_diff(message, differ_factory, actual)
+        diff = diffs(differ_factory, actual)
+        message = "#{message}\n#{diff}" unless diff.empty?
+        message
+      end
+
+    private
+
+      def self.diff_label_for(matcher)
+        "Diff for (#{truncated(matcher.description)}):"
```

Comment:
I forgot to mention this earlier...`description` is optional (but encouraged), too. See here for the protocol docs:

http://www.rubydoc.info/github/rspec/rspec-expectations/RSpec/Matchers/MatcherProtocol

Thus, we need to handle matchers which do not have a description.  Elsewhere we've handled it with a simple rescue:

https://github.com/rspec/rspec-expectations/blob/45ff9faa26ea912d6b13d4044e5691b894d9fcc5/lib/rspec/expectations/expectation_target.rb#L112-L116

The thinking is that in 99% of cases the matcher will have `description`, so while `respond_to?` is more performant that allowing an exception to be raised, that only comes into play 1% of the time and the rest of the time it's faster to not check `respond_to?`.

Honestly, I don't particularly care which route you go to deal with it (respond to or exception rescue).

It'd also be nice to have specs for matchers that don't support `diffable?` and `description` since it's an easy thing to forget.